<a href="https://colab.research.google.com/github/eilselx/MACHINE-LEARNING-PROJECT---LIFE-EXPECTANCY-ESTIMATOR-/blob/main/MACHINE_LEARNING_PROJECT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

The First Ever Machine Learning Project!                 


Problem Statement: Can machine learning accurately predict a country's life expectancy using demographic, healthcare, economic, and public health indicators?

Section 1 - By Lavanya & Sikhin:

Data Cleaning, Preprocessing, Exploratory Data Analysis (EDA), Data Visualization


In [1]:
# Cloning repository and changing directory
import os
repo_name = 'MACHINE-LEARNING-PROJECT---LIFE-EXPECTANCY-ESTIMATOR-'

# Check if the directory already exists
if not os.path.exists(repo_name):
    print(f"Cloning {repo_name} repository...")
    !git clone https://github.com/eilselx/{repo_name}.git
    print(f"Repository {repo_name} cloned.")
else:
    print(f"Repository {repo_name} already exists. Skipping clone.")

# Change into the repository directory
if os.path.exists(repo_name):
    %cd {repo_name}
    print(f"Changed directory to: {os.getcwd()}")
else:
    print(f"Error: {repo_name} directory not found after cloning attempt.")

Cloning MACHINE-LEARNING-PROJECT---LIFE-EXPECTANCY-ESTIMATOR- repository...
Cloning into 'MACHINE-LEARNING-PROJECT---LIFE-EXPECTANCY-ESTIMATOR-'...
remote: Enumerating objects: 29, done.
remote: Counting objects: 100% (29/29), done.
remote: Compressing objects: 100% (22/22), done.
remote: Total 29 (delta 5), reused 4 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (29/29), 1.32 MiB | 9.06 MiB/s, done.
Resolving deltas: 100% (5/5), done.
Repository MACHINE-LEARNING-PROJECT---LIFE-EXPECTANCY-ESTIMATOR- cloned.
/content/MACHINE-LEARNING-PROJECT---LIFE-EXPECTANCY-ESTIMATOR-
Changed directory to: /content/MACHINE-LEARNING-PROJECT---LIFE-EXPECTANCY-ESTIMATOR-


In [2]:
#import dataset
import pandas as pd
import zipfile
import os

# Define the path to the zip file
zip_file_path = "data/archive(1).zip"
# Define the directory to extract to (can be the current directory or a subfolder)
extract_dir = "data"

# Unzip the file if it hasn't been unzipped already
if os.path.exists(zip_file_path) and not os.path.exists(os.path.join(extract_dir, 'health_indicators.csv')):
    print(f"Unzipping {zip_file_path}...")
    with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
        zip_ref.extractall(extract_dir)
    print("Zip file unzipped.")

# List files in the 'data' directory to confirm extraction
print("Files in data directory after (attempted) unzipping:")
for item in os.listdir(extract_dir):
    print(item)

# Now read the CSV file from the extracted location
df = pd.read_csv(os.path.join(extract_dir, "health_indicators.csv"))
print("Dataset loaded successfully. First 5 rows:")
print(df.head())
df.shape
df.info()


Unzipping data/archive(1).zip...
Zip file unzipped.
Files in data directory after (attempted) unzipping:
archive(1).zip
README.md
country_latest.csv
indicator_metadata.csv
cleaned_health_indicators_data.csv
health_indicators.csv
Dataset loaded successfully. First 5 rows:
  country_code country_name                      region income_level  year  \
0          ABW        Aruba  Latin America & Caribbean   High income  2000   
1          ABW        Aruba  Latin America & Caribbean   High income  2001   
2          ABW        Aruba  Latin America & Caribbean   High income  2002   
3          ABW        Aruba  Latin America & Caribbean   High income  2003   
4          ABW        Aruba  Latin America & Caribbean   High income  2004   

   population  population_growth  urban_pct  life_expectancy  \
0     90588.0           1.030817  65.354550           72.939   
1     91439.0           0.935033  65.335114           73.044   
2     92074.0           0.692052  65.282069           73.135   
3  

Cleaning report of the Import cell:
1. No. of Rows: 5275
2. No. of columns: 48
3. Data Types: 43 floats, 1 int, 4 objects


In [3]:
#working on missing values
df.isnull().sum()
print(df.isnull().sum())

#percentage of missing values for the columns
missing_percentage = (df.isnull().sum() / len(df)) * 100
print(missing_percentage)

#Table format
missing_table = pd.DataFrame({
    'Missing Count': df.isnull().sum(),
    'Missing Percentage': missing_percentage
})

#I need to see how many are more than 50 percent missing value (I guess that cannot be used in the model as it won't be accurate to predict)
missing_table = missing_table[missing_table['Missing Percentage'] > 50]
print(missing_table)


country_code                        0
country_name                        0
region                              0
income_level                        0
year                                0
population                          0
population_growth                   1
urban_pct                           0
life_expectancy                     0
life_expectancy_male                0
life_expectancy_female              0
death_rate                          0
birth_rate                          0
health_expenditure_pct_gdp        761
health_expenditure_per_capita     764
physicians_per_1000              2444
hospital_beds_per_1000           2377
nurses_per_1000                  2443
under5_mortality                  425
neonatal_mortality                425
maternal_mortality                667
tb_incidence                      218
hiv_incidence                    1661
communicable_death_pct           4177
noncommunicable_death_pct        4177
smoking_male                     3971
smoking_fema

Specifically analysing missing values (imp points):

- No missing values in life expectancy columns - all 3
- All values for obesity % is missing
- 12 coulmns don't have more than 50 % data
- All important indicators that can have great impacts of life expectancy has more than 50 % data missing - such as NCD death %, smoking, CDs %, handwash %, prenatal care% & poverty rates


In [4]:
#checking for duplication, types of data
df.duplicated().sum()
df.dtypes.value_counts()
print(df.select_dtypes(include='object').columns)
df.select_dtypes(include='int64').columns

Index(['country_code', 'country_name', 'region', 'income_level'], dtype='object')


Index(['year'], dtype='object')

So,
43 columns are floats - ready for analysis - indicators

4 columns are objects  - country_name, country_code, region, income level (not sure which to keep and drop)

1 column is an integer that contains years

In [5]:
#inspecting the columbs
print(df["region"].value_counts())
print(df["income_level"].value_counts())
print(df["country_name"].value_counts())
print(df["country_code"].value_counts())
print(df["year"].value_counts())

region
Europe & Central Asia                                1400
Sub-Saharan Africa                                   1200
Latin America & Caribbean                            1050
East Asia & Pacific                                   875
Middle East, North Africa, Afghanistan & Pakistan     525
South Asia                                            150
North America                                          75
Name: count, dtype: int64
income_level
High income            2025
Upper middle income    1350
Lower middle income    1225
Low income              625
Not classified           50
Name: count, dtype: int64
country_name
Aruba           25
Afghanistan     25
Angola          25
Albania         25
Andorra         25
                ..
Kosovo          25
Yemen, Rep.     25
South Africa    25
Zambia          25
Zimbabwe        25
Name: count, Length: 211, dtype: int64
country_code
ABW    25
AFG    25
AGO    25
ALB    25
AND    25
       ..
XKX    25
YEM    25
ZAF    25
ZMB    25
ZWE    2

So, 211 countries are observed throgh 25 years.

7 regions:

Europe & Central Asia                               
Sub-Saharan Africa                                   
Latin America & Caribbean                            
East Asia & Pacific                                  
Middle East, North Africa, Afghanistan & Pakistan     
South Asia                                            
North America

5 Income Levels: (Can we code this as 5,4,3,2,1 ?)

High Income,
Upper Middle Income,
Lower Middle Income,
Low Income,
Not Classified.


In [6]:
#checking all mabout the LE as it's main predictor
df.life_expectancy.describe()
life_expectancy_comp = df[["life_expectancy", "life_expectancy_male", "life_expectancy_female"]]
print(life_expectancy_comp.describe())
df.describe()
(df.select_dtypes(include=['float64','int64']) < 0).sum()
df['population_growth'].describe()


       life_expectancy  life_expectancy_male  life_expectancy_female
count      5275.000000           5275.000000             5275.000000
mean         70.748244             68.215760               73.394403
std           8.701547              8.471691                9.066795
min          14.665000             12.383000               20.006000
25%          65.356000             62.722000               67.976000
50%          72.442000             69.217000               75.689000
75%          77.197000             74.500000               80.115000
max          86.497000             84.560000               88.626000


,population_growth
count,5274.000000
mean,1.280475
std,1.618914
min,-11.356645
25%,0.297465
50%,1.175030
75%,2.210221
max,21.700343


- Nothing missing in the main predictor value: Life expectancy
- there are no impossible values as such seen. Only population growth showed 883 negative values which can happen as population growth can decline. both -11 % adn 21.7 %seems possible.



In [7]:
#deleting columns with more than 80 % missing values, filling in those with lesser missing values.
df.isnull().sum()
missing_percentage = (df.isnull().sum() / len(df)) * 100
dropped_columns = missing_percentage[missing_percentage > 80].index
len(dropped_columns)
dropped_columns
df = df.drop(columns=dropped_columns)
missing_after_drop = df.isnull().sum()
missing_after_drop[missing_after_drop > 0]

,0
population_growth,1
health_expenditure_pct_gdp,761
health_expenditure_per_capita,764
physicians_per_1000,2444
hospital_beds_per_1000,2377
nurses_per_1000,2443
under5_mortality,425
neonatal_mortality,425
maternal_mortality,667
tb_incidence,218


- All columns with more than 80 percent missing values are dropped - Totall 6 columns = obesity_pct', 'diabetes_pct', 'stunting_pct', 'wasting_pct', 'overweight_children_pct', 'prenatal_care_pct'

In [8]:
#filling in the missing values
df[df.columns[df.isnull().any()]].dtypes
numeric_columns = df.select_dtypes(include=['float64', 'int64']).columns
df[numeric_columns] = df[numeric_columns].fillna(
    df[numeric_columns].median())
df.isnull().sum().sum()
df.to_csv('data/cleaned_health_indicators_data.csv', index=False)


- All the missing values are filled using median imputation.
- All check points are done for cleaning data. EDA needs to be done now.
- Saved as "cleaned health indicators data"